File: src/lib/clients/release-file-removal.ts
/**
 * Physically remove a deleted release's files — and only its own files.
 *
 * WHY THIS EXISTS
 * ---------------
 * "Delete + files" on the Downloads page promised to remove the download from
 * disk and did not. The built-in engine's `deleteTorrent` only reliably deleted
 * bytes when a *live* WebTorrent handle was still in memory
 * (`destroy({ destroyStore: true })`). For a rehydrated row with no live handle
 * — the ordinary case after a restart, and the exact shape of the seeded
 * `.e2e-instant-play` fixtures — it deleted the database row and then merely
 * pruned *empty* parent folders. The release folder was not empty (it still held
 * the media file plus sidecar junk), so nothing was pruned and every byte
 * survived while the row, and the UI row, vanished.
 *
 * And even the live path left litter behind: WebTorrent's store destroy removes
 * the files the torrent wrote, not the `.torrentflow` marker folder, the
 * `Featurettes` extras, the "Torrent Downloaded From ….txt" tracker spam, or a
 * half-written `.part`.
 *
 * THE RULE
 * --------
 * Delete exactly what this torrent owns, proven from its *recorded* absolute
 * file paths (`EngineTorrent.verifiedFilesJson`), never from a path a client
 * merely claims. Then, when the torrent created its own folder for that release,
 * remove that whole folder — junk included — but only after proving:
 *
 *   1. the folder is strictly *inside* the download root (never the root, never
 *      an ancestor of it);
 *   2. the folder is inside-or-equal the torrent's own recorded save path
 *      (never a directory above it);
 *   3. no *other* torrent still holds anything inside that folder — otherwise it
 *      is a shared category/show directory and only this release's own files may
 *      go.
 *
 * That is what keeps a shared `…/TV/` root, or a `Rick and Morty/` folder that
 * two episodes share, safe: the folder is removed only when it is demonstrably
 * this one release's own container.
 *
 * PURE + INJECTABLE FS
 * --------------------
 * `planReleaseRemoval` is pure so the containment rule can be checked as a
 * table. `executeReleaseRemoval` takes an injectable {@link RemovalFs}, mirroring
 * `add-allocation.ts`, so the unlink/rmtree behaviour is testable without a disk
 * and provable against a real one from a probe script.
 */
import fs from "node:fs";
import path from "node:path";

import { isPathInsideOrEqual } from "./prune-empty-parents";

export interface ReleaseRemovalRequest {
  /** Absolute file paths recorded as belonging to this torrent. */
  ownedFiles: readonly string[];
  /** The torrent's recorded save directory (its leaf save path). */
  savePath?: string | null;
  /** Download root. Nothing at or above this is ever removed. */
  baseRoot?: string | null;
  /**
   * Absolute paths (files, or save directories) still owned by OTHER torrents.
   * If any of these lives inside the candidate folder, the folder is shared and
   * is not removed — only this release's own files are.
   */
  otherPaths?: readonly string[];
}

export interface ReleaseRemovalPlan {
  /** Absolute file paths to unlink. Each proven strictly inside `baseRoot`. */
  files: string[];
  /** The torrent's own folder, safe to remove recursively — or null. */
  folder: string | null;
  /** Owned paths refused for being outside `baseRoot`. Never touched. */
  refusedOutside: string[];
  /** Why {@link folder} is null, when it is. Diagnostic only. */
  folderReason: string | null;
}

/** Injectable filesystem seam so the rule can be tested without a disk. */
export interface RemovalFs {
  /** File size in bytes, or null when the path is absent or not a file. */
  statSize: (file: string) => number | null;
  /** Remove a single file. Best-effort; may throw, callers catch. */
  removeFile: (file: string) => void;
  /** Recursively remove a directory and everything under it. */
  removeTree: (dir: string) => void;
}

export interface ReleaseRemovalResult {
  removed: string[];
  freedBytes: number;
  folderRemoved: string | null;
}

function norm(p: string): string {
  return process.platform === "win32"
    ? path.resolve(p).replace(/\//g, "\\").toLowerCase()
    : path.resolve(p);
}

/** True when `child` is inside `root`, but not `root` itself. */
export function isStrictlyInside(child: string, root: string): boolean {
  return isPathInsideOrEqual(child, root) && norm(child) !== norm(root);
}

/**
 * The deepest directory that contains every one of `files`.
 *
 * For a single file this is its parent. For a multi-file torrent it is the
 * folder the torrent created for the release — which is exactly the thing we
 * want to remove when it is the torrent's own. Case-insensitive on Windows, but
 * the original casing of the first path is preserved in the result.
 */
export function commonAncestorDir(files: readonly string[]): string | null {
  const dirs = files
    .map((f) => f?.trim())
    .filter((f): f is string => Boolean(f))
    .map((f) => path.dirname(path.resolve(f)));
  if (dirs.length === 0) return null;

  const split = (p: string) => p.split(/[\\/]+/);
  const cmp = (a: string, b: string) =>
    process.platform === "win32"
      ? a.toLowerCase() === b.toLowerCase()
      : a === b;

  let prefix = split(dirs[0]);
  for (const dir of dirs.slice(1)) {
    const parts = split(dir);
    let i = 0;
    while (i < prefix.length && i < parts.length && cmp(prefix[i], parts[i])) i++;
    prefix = prefix.slice(0, i);
    if (prefix.length === 0) return null;
  }
  const joined = prefix.join(path.sep);
  // A bare drive letter ("C:") joins to "C:"; resolve normalises it to a root
  // path, which strict-inside checks will then correctly reject.
  return path.resolve(joined);
}

/**
 * Decide what deleting this release is allowed to remove from disk.
 *
 * Total and conservative: every refusal costs at most a leftover empty folder,
 * every wrong permission costs the user files. When in doubt, only the recorded
 * files go and the folder is left for the empty-parent prune to reconsider.
 */
export function planReleaseRemoval(
  req: ReleaseRemovalRequest,
): ReleaseRemovalPlan {
  const baseRoot = req.baseRoot?.trim() || null;
  const savePath = req.savePath?.trim() || null;

  const files: string[] = [];
  const refusedOutside: string[] = [];
  const seen = new Set<string>();

  for (const raw of req.ownedFiles) {
    const p = raw?.trim();
    if (!p) continue;
    const resolved = path.resolve(p);
    const key = norm(resolved);
    if (seen.has(key)) continue;
    seen.add(key);
    // No base root means we cannot prove containment, so we may not delete.
    if (!baseRoot || !isStrictlyInside(resolved, baseRoot)) {
      refusedOutside.push(resolved);
      continue;
    }
    files.push(resolved);
  }

  const plan: ReleaseRemovalPlan = {
    files,
    folder: null,
    refusedOutside,
    folderReason: null,
  };

  if (!baseRoot) {
    plan.folderReason = "no download root to bound the delete";
    return plan;
  }
  if (files.length === 0) {
    // Nothing recorded that we could prove owns a folder. Leave the folder to
    // the empty-parent prune rather than guessing at a directory to remove.
    plan.folderReason = "no recorded files to prove folder ownership";
    return plan;
  }

  const candidate = commonAncestorDir(files);
  if (!candidate) {
    plan.folderReason = "files share no common folder";
    return plan;
  }
  if (!isStrictlyInside(candidate, baseRoot)) {
    plan.folderReason = "folder is the download root or above it";
    return plan;
  }
  if (savePath && !isPathInsideOrEqual(candidate, savePath)) {
    plan.folderReason = "folder is above the recorded save path";
    return plan;
  }

  const others = (req.otherPaths ?? [])
    .map((p) => p?.trim())
    .filter((p): p is string => Boolean(p));
  for (const other of others) {
    if (isPathInsideOrEqual(other, candidate)) {
      plan.folderReason = "folder is shared with another download";
      return plan;
    }
  }

  plan.folder = candidate;
  return plan;
}

/**
 * Carry out a {@link ReleaseRemovalPlan}. Never throws.
 *
 * Files first (so a locked media file is still attempted before its folder),
 * then, if the plan proved a folder is this release's own, the whole folder —
 * which is what takes `Featurettes`, `.torrentflow`, the "Torrent Downloaded
 * From …" text file and any `.part` with it. A file that cannot be removed is
 * simply not counted as freed; cleanup on a delete path must not become a throw.
 */
export function executeReleaseRemoval(
  plan: ReleaseRemovalPlan,
  io: RemovalFs = realRemovalFs,
): ReleaseRemovalResult {
  const removed: string[] = [];
  let freedBytes = 0;

  for (const file of plan.files) {
    try {
      const size = io.statSize(file);
      io.removeFile(file);
      if (typeof size === "number" && size > 0) freedBytes += size;
      removed.push(file);
    } catch {
      /* locked or already gone — not freed, not fatal */
    }
    // A sibling half-written chunk file is this file's own and never anyone's
    // media, so it is safe to remove even outside a proven-own folder.
    for (const suffix of [".part", ".!qB"]) {
      try {
        io.removeFile(`${file}${suffix}`);
      } catch {
        /* absent or locked */
      }
    }
  }

  let folderRemoved: string | null = null;
  if (plan.folder) {
    try {
      io.removeTree(plan.folder);
      folderRemoved = plan.folder;
    } catch {
      /* best-effort: an empty-parent prune runs afterwards regardless */
    }
  }

  return { removed, freedBytes, folderRemoved };
}

export const realRemovalFs: RemovalFs = {
  statSize: (file) => {
    const s = fs.statSync(file, { throwIfNoEntry: false });
    return s && s.isFile() ? s.size : null;
  },
  removeFile: (file) => {
    fs.rmSync(file, { force: true });
  },
  removeTree: (dir) => {
    fs.rmSync(dir, { recursive: true, force: true });
  },
};

File: server/TorrentFlow.Engine/Storage/ReleaseFileRemoval.cs
namespace TorrentFlow.Engine.Storage;

/// <summary>What deleting one release may remove from disk (port of <c>release-file-removal.ts</c>).</summary>
internal sealed record ReleaseRemovalPlan(IReadOnlyList<string> Files, string? Folder, string? FolderReason);

/// <summary>
/// Physically removes a deleted release's files and, when the release created its own folder, that whole folder —
/// sidecar junk (posters, tracker spam, <c>.part</c> files) included — then prunes the empty folders it leaves behind,
/// down and up, never at or above the download root. A shared category or show folder is never removed: any path
/// another transfer still records inside it keeps it.
/// </summary>
internal static class ReleaseFileRemoval
{
    private static StringComparison Comparison => OperatingSystem.IsWindows() ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;

    public static bool IsInsideOrEqual(string child, string root)
    {
        var c = Path.TrimEndingDirectorySeparator(Path.GetFullPath(child));
        var r = Path.TrimEndingDirectorySeparator(Path.GetFullPath(root));
        return string.Equals(c, r, Comparison) || c.StartsWith(r + Path.DirectorySeparatorChar, Comparison);
    }

    public static bool IsStrictlyInside(string child, string root) =>
        IsInsideOrEqual(child, root) && !string.Equals(Path.TrimEndingDirectorySeparator(Path.GetFullPath(child)),
            Path.TrimEndingDirectorySeparator(Path.GetFullPath(root)), Comparison);

    /// <summary>The deepest directory holding every file, or null when they share none.</summary>
    public static string? CommonAncestorDir(IEnumerable<string> files)
    {
        List<string>? prefix = null;
        foreach (var file in files)
        {
            var dir = Path.GetDirectoryName(Path.GetFullPath(file));
            if (dir is null) return null;
            var parts = dir.Split(Path.DirectorySeparatorChar, StringSplitOptions.None).ToList();
            if (prefix is null) { prefix = parts; continue; }
            var i = 0;
            while (i < prefix.Count && i < parts.Count && string.Equals(prefix[i], parts[i], Comparison)) i++;
            prefix = prefix.Take(i).ToList();
            if (prefix.Count == 0) return null;
        }
        if (prefix is null || prefix.Count == 0) return null;
        var joined = string.Join(Path.DirectorySeparatorChar, prefix);
        return Path.GetFullPath(joined.EndsWith(':') ? joined + Path.DirectorySeparatorChar : joined);
    }

    /// <summary>
    /// Pure. Recorded files inside the save path or the download root may go. The release's folder goes too when it
    /// is strictly inside the root, inside-or-equal the save path, holds nothing another transfer records, and is the
    /// release's own container: strictly below the save path, or at least two levels below the root. The last rule is
    /// stricter than Next, whose plan would take a whole category folder (<c>Other/</c>) with a single-file release.
    /// </summary>
    public static ReleaseRemovalPlan Plan(IEnumerable<string> ownedFiles, string? savePath, string? baseRoot, IEnumerable<string> otherPaths)
    {
        savePath = string.IsNullOrWhiteSpace(savePath) ? null : Path.GetFullPath(savePath.Trim());
        baseRoot = string.IsNullOrWhiteSpace(baseRoot) ? null : Path.GetFullPath(baseRoot.Trim());
        var files = new List<string>();
        var seen = new HashSet<string>(OperatingSystem.IsWindows() ? StringComparer.OrdinalIgnoreCase : StringComparer.Ordinal);
        foreach (var raw in ownedFiles)
        {
            if (string.IsNullOrWhiteSpace(raw)) continue;
            var full = Path.GetFullPath(raw.Trim());
            if (!seen.Add(full)) continue;
            if ((savePath is not null && IsStrictlyInside(full, savePath)) || (baseRoot is not null && IsStrictlyInside(full, baseRoot)))
                files.Add(full);
        }

        if (baseRoot is null) return new(files, null, "no download root to bound the delete");
        if (files.Count == 0) return new(files, null, "no recorded files to prove folder ownership");
        var candidate = CommonAncestorDir(files);
        if (candidate is null) return new(files, null, "files share no common folder");
        if (!IsStrictlyInside(candidate, baseRoot)) return new(files, null, "folder is the download root or above it");
        if (savePath is not null && !IsInsideOrEqual(candidate, savePath)) return new(files, null, "folder is above the recorded save path");
        var depth = Path.GetRelativePath(baseRoot, candidate).Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries).Length;
        if (depth < 2 && !(savePath is not null && IsStrictlyInside(candidate, savePath)))
            return new(files, null, "folder is a category folder");
        foreach (var other in otherPaths)
            if (!string.IsNullOrWhiteSpace(other) && IsInsideOrEqual(other.Trim(), candidate))
                return new(files, null, "folder is shared with another download");
        return new(files, candidate, null);
    }

    /// <summary>Never throws: files first (and their half-written siblings), then the proven-own folder.</summary>
    public static void Execute(ReleaseRemovalPlan plan)
    {
        foreach (var file in plan.Files)
        {
            TryDeleteFile(file);
            TryDeleteFile(file + ".part");
            TryDeleteFile(file + ".!qB");
        }
        if (plan.Folder is { } folder)
        {
            try
            {
                if (Directory.Exists(folder) && !new DirectoryInfo(folder).Attributes.HasFlag(FileAttributes.ReparsePoint))
                    Directory.Delete(folder, recursive: true);
            }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException) { }
        }
    }

    /// <summary>Removes empty directories below <paramref name="start"/>, deepest first; links are never followed.</summary>
    public static void PruneEmptyDescendants(string? start, string? stopAt, int maxDepth = 8)
    {
        if (string.IsNullOrWhiteSpace(start) || string.IsNullOrWhiteSpace(stopAt)) return;
        var root = Path.GetFullPath(stopAt);
        var from = Path.GetFullPath(start);
        if (!IsInsideOrEqual(from, root) || !Directory.Exists(from)) return;
        var visits = 0;
        void Walk(string dir, int depth)
        {
            if (depth > maxDepth || visits++ > 2000) return;
            IEnumerable<DirectoryInfo> children;
            try { children = new DirectoryInfo(dir).EnumerateDirectories().ToList(); }
            catch (Exception ex) when (ex is IOException or UnauthorizedAccessException) { return; }
            foreach (var child in children)
            {
                if (child.Attributes.HasFlag(FileAttributes.ReparsePoint)) continue;
                Walk(child.FullName, depth + 1);
                try { if (!child.EnumerateFileSystemInfos().Any()) child.Delete(); }
                catch (Exception ex) when (ex is IOException or UnauthorizedAccessException) { }
            }
        }
        Walk(from, 0);
    }

    private static void TryDeleteFile(string path)
    {
        try { if (File.Exists(path)) File.Delete(path); }
        catch (IOException) { }
        catch (UnauthorizedAccessException) { }
    }
}
